driver-only ECC: curve acceleration macros

valeriosetti opened the pull request:
There is an [ongoing discussion](https://github.com/Mbed-TLS/mbedtls/pull/8075#issuecomment-1678825516) about the scope of this PR. Once clarified I will update this description accordingly.

~Depends on~ Supersedes #8033 
Resolves #8016, resolves #8014.

## PR checklist

- [ ] **changelog** deferred to #7757
- [ ] **backport** not required since it's an enhancement
- [ ] **tests** not required no new code to be tested

daverodgman commented:
@valeriosetti writes:

For those unfamiliar with this task and its ancestor ([issue 8014](https://github.com/Mbed-TLS/mbedtls/issues/8014)) the idea is to:

- ensure that the curves supported on the PSA side are always a super-set of the builtin ones (task 8014)
- enable builtin algs as long as there is some builtin curve supported and the corresponding PSA_WANT_ALG is enabled (task 8016)
- this requires changes to some test components in all.sh: component_test_psa_crypto_config_accel_ecc_ecp_light_only, component_test_psa_crypto_config_accel_ecc_no_ecp_at_all and component_test_psa_crypto_config_accel_ecc_no_bignum

I almost addressed all the requirements Manuel wrote in 8016, but I'm facing a problem is with the component_test_psa_crypto_config_accel_ecc_ecp_light_only :

- ECP_LIGHT is auto enabled in build_info.h when some other symbol is defined, like MBEDTLS_PK_PARSE_EC_EXTENDED and MBEDTLS_PK_PARSE_EC_COMPRESSED
  - this is required because this support is missing in PSA
  - of course to make this support to work I need to keep MBEDTLS_ECP_DP symbols enabled
- Unfortunately, with changes requested in 8016, as soon as I re-enable builtin curve support (any of MBEDTLS_ECP_DP ) then all the "undesired" modules ( ECDSA_C , ECDH_C , ECP_C ) are included into the build
  - IMO this makes the test component itself useless because we would actually have full ECP support also on the builtin side :confused:

Long story short: test components which do not require any builtin support of ECP (component_test_psa_crypto_config_accel_ecc_no_ecp_at_all  and component_test_psa_crypto_config_accel_ecc_no_bignum ) work just fine. Mixed one (component_test_psa_crypto_config_accel_ecc_ecp_light_only ) does not.

Do you have any suggestion on how to overcome this problem?

valeriosetti commented:
I don't have access to the Internal CI, but since the Open component fully passed, I think we can ignore that and consider the PR finally ready for review

gilles-peskine-arm commented:
There was maintenance on the armcc license server used by the internal CI. Since the armcc builds passed on OpenCI, this is indeed a pass.

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -2487,26 +2520,28 @@ component_test_psa_crypto_config_accel_ecc_ecp_light_only () {
     # Use the same config as reference, only without built-in EC algs
     config_psa_crypto_config_ecp_ligh_only 1
 
+    # Do not disable builtin curves because that support is rquired for:
```
```suggestion
    # Do not disable builtin curves because that support is required for:
```

gilles-peskine-arm commented on tests/ssl-opt.sh:
```diff
@@ -11300,8 +11303,10 @@ run_test    "TLS 1.3: Test gnutls tls1_3 feature" \
 requires_ciphersuite_enabled TLS1-3-CHACHA20-POLY1305-SHA256
-requires_config_enabled MBEDTLS_ECP_DP_CURVE25519_ENABLED
-requires_config_enabled MBEDTLS_ECP_DP_SECP256R1_ENABLED
+requires_any_configs_enabled "MBEDTLS_ECP_DP_CURVE25519_ENABLED
+                              PSA_WANT_ECC_MONTGOMERY_255"
+requires_any_configs_enabled "MBEDTLS_ECP_DP_SECP256R1_ENABLED \
+                              PSA_WANT_ECC_SECP_R1_256"
```
As in `ssl_misc.h`, this is TLS 1.3 which is PSA-only so
```suggestion
requires_config_enabled PSA_WANT_ECC_MONTGOMERY_255
requires_config_enabled PSA_WANT_ECC_SECP_R1_256
```

gilles-peskine-arm commented on tests/suites/test_suite_pk.data:
```diff
@@ -13,135 +13,135 @@ depends_on:MBEDTLS_RSA_C:MBEDTLS_GENPRIME
 pk_utils:MBEDTLS_PK_RSA:MBEDTLS_RSA_GEN_KEY_MIN_BITS:MBEDTLS_RSA_GEN_KEY_MIN_BITS:(MBEDTLS_RSA_GEN_KEY_MIN_BITS /8):"RSA"
```
There's one occurrence of `MBEDTLS_ECP_DP_SECP384R1_ENABLED` in `test_suite_psa_crypto.data` which should be replaced as well.

gilles-peskine-arm commented on include/mbedtls/build_info.h:
```diff
@@ -221,6 +221,47 @@
 
+/* Helpers to state that each key is supported either on the builtin or PSA side. */
+#if defined(MBEDTLS_ECP_DP_SECP521R1_ENABLED) || defined(PSA_WANT_ECC_SECP_R1_521)
+#define MBEDTLS_HAVE_SECP521R1
```
This name feels like it's missing a namespace. `MBEDTLS_ECP_HAVE_SECP521R1`, I guess, since the scope is ECP-light?

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -2318,11 +2357,9 @@ component_test_psa_crypto_config_accel_ecdh () {
 
```
I don't think this satisfies the test requirements in https://github.com/Mbed-TLS/mbedtls/issues/8016. Specifically:

> at least test one case where there's one accelerated curve and one built-in curve, one case where ECDH is accelerated but not ECDSA, and one case where it's the other way round.

for which [Manuel proposed a strategy](https://github.com/Mbed-TLS/mbedtls/issues/8016#issuecomment-1663446576).

valeriosetti replied:
I thought that for the 1st part ("at least test one case where there's one accelerated curve and one built-in curve") the following components were enough:
- `test_psa_crypto_config_accel_all_curves_except_p192`
- `test_psa_crypto_config_accel_all_curves_except_x25519`

and that for the 2nd one ("one case where ECDH is accelerated but not ECDSA, and one case where it's the other way round") these were enough:
- `test_psa_crypto_config_accel_ecdh`
- `test_psa_crypto_config_accel_ecdsa`

Or you were asking for a more detailed scenario in which both 1st and 2nd case were mixed?
I mean, something like:
- `accel_all_curves_except_p192` + accelerate ECDH only
- `accel_all_curves_except_p192` + accelerate ECDSA only
- `accel_all_curves_except_x25519` + accelerate ECDH only
- `accel_all_curves_except_x25519` + accelerate ECDSA only
?

gilles-peskine-arm replied:
Thanks for the breakdown, I might have missed some of those in my review. I'll re-check.

gilles-peskine-arm replied:
Having re-checked, I'm satisfied that the existing components provide enough coverage. Sorry for the noise.

valeriosetti replied:
No problem at all! Thanks for the review :)

gilles-peskine-arm reviewed (changes requested):
ac8614a2e707e4cbcd5ac125295685d38e08bb7e Looks mostly good, just missing some testing adjustments and a few minor points.

valeriosetti commented on include/mbedtls/check_config.h:
```diff
@@ -66,6 +66,71 @@
+#if defined(MBEDTLS_SOME_BUILTIN_EC) && \
+    !(defined(MBEDTLS_ECDSA_C) || defined(MBEDTLS_ECDSA_DETERMINISTIC) || \
+      defined(MBEDTLS_ECDH_C) || defined(MBEDTLS_ECJPAKE_C))
+#error "Some builtin curve is included, but there is no algorighm using it"
```
Should I replace it with a `#warning` then? Or is it better to remove the check entirely?

Another (perhaps weird) option that comes into my mind is to create a new build symbol like `MBEDTLS_CUSTOM_EC_USAGE` and add it to this list. It wouldn't be used anywhere in the lib/test/program codes though, unless we design some specific example for that...

Wdyt?

valeriosetti replied:
Sorry @gilles-peskine-arm to resume this discussion again, but I think I got to a point in which I need your feedback.

I'm rebasing this PR after the merge of #8159 and I realized that previous to this rebase I was basically not using the new internal symbol `MBEDTLS_SOME_BUILTIN_EC` anywhere in the code. I was only defining it. I guess this does not satisfy Manuel's original requirement in #8016 (2nd part of the sentence after the "OR"):

> check that flag when deciding whether to enable ECDH_C, ECDSA_C and ECJPAKE_C: we'll enable the built-in implementation if there's no accelerator for this alg OR if the flag is set (which is the new part)

I come up with my proposal to add a check in `check_config.h` after [I realized](https://github.com/Mbed-TLS/mbedtls/pull/8075#issuecomment-1678825516) that I were implementing things as Manuel asked, then some test components were failing:
- `component_test_psa_crypto_config_accel_ecdsa`
- `component_test_psa_crypto_config_accel_ecdh`
- `component_test_psa_crypto_config_accel_ecc_ecp_light_only`
- `component_test_psa_crypto_config_accel_all_curves_except_p192`

Let's pick for example `component_test_psa_crypto_config_accel_ecc_ecp_light_only`. There we accelerate all EC algs, but at the same time we keep builtin curves because they are needed for `ECP_LIGHT` features like `MBEDTLS_PK_PARSE_EC_EXTENDED`. Unfortunately in that test component we check that:
```
    not grep mbedtls_ecdsa_ library/ecdsa.o
    not grep mbedtls_ecdh_ library/ecdh.o
    not grep mbedtls_ecjpake_ library/ecjpake.o
    not grep mbedtls_ecp_mul library/ecp.o
```
and this of course fails because in `config_adjust_legacy_from_psa.h` we have something like:
```
#if defined(PSA_WANT_ALG_ECDH)
#if !defined(MBEDTLS_PSA_ACCEL_ALG_ECDH) || defined(MBEDTLS_SOME_BUILTIN_EC)
#define MBEDTLS_PSA_BUILTIN_ALG_ECDH 1
#define MBEDTLS_ECDH_C
#define MBEDTLS_ECP_C
#define MBEDTLS_BIGNUM_C
#endif /* !MBEDTLS_PSA_ACCEL_ALG_ECDH */
#endif /* PSA_WANT_ALG_ECDH */
```
which brings the builtin implementation in even though we have ECDH accelerated.

How do you suggest to overcome this problem?

valeriosetti replied:
This commit https://github.com/Mbed-TLS/mbedtls/pull/8075/commits/7ef50aeb3886b64f15c0d17024cf11208b1fe612 is the best compromise I can think of. It matches Manuel's requirement and at the same time should make the CI happy.
I'm not super happy with it because it partially re-uses the same symbols that are used to enable `ECP_LIGHT`. It might be more elegant to use `ECP_LIGHT` instead here.
Anyway, is this a reasonable solution to you?

gilles-peskine-arm replied:
Checking again, I think you're right that 55efa0bc693ead13063609be5e497cc4ef23308d doesn't meet the requirements in #8016. But neither does d60abc8dfca9c4815678798035e56d22d8eeb6d1.

See https://github.com/Mbed-TLS/mbedtls/pull/8075#discussion_r1319836847

valeriosetti commented:
Note: While solving recent comments I also rebased initial commits in order to follow the development of #8033

gilles-peskine-arm reviewed (changes requested):
LGTM for the result, but there's a meaningless commit that should be squashed.

valeriosetti commented on include/mbedtls/config_psa.h:
```diff
@@ -102,6 +102,7 @@
 #define PSA_WANT_ALG_SHA3_256 1
 #define PSA_WANT_ALG_SHA3_384 1
 #define PSA_WANT_ALG_SHA3_512 1
+#endif
```
Sure! However that commit belong to the preceding PR (#8033), so I will fix that one and rebase this one on top of it.

Since we are talking about it and you already reviewed this one, can you please take a look also to #8033? Commits there are the initial ones of this PR, so there won't be anything new...

gilles-peskine-arm replied:
#8033 needs this squashing as well.

valeriosetti replied:
Yes, correct, indeed I rebased/modified #8033 last friday, so this unnecessary commit is no more listed in its history

tom-daubney-arm commented on tests/scripts/all.sh:
```diff
@@ -887,6 +887,47 @@ helper_libtestdriver1_adjust_config() {
 }
 
+# When called with no parameter this function disables all builtin curves.
+# The function optionally accepts 1 parameter with the space-separated list
```
```suggestion
# The function optionally accepts 1 parameter: a space-separated list
```

tom-daubney-arm reviewed (dismissed):
This LGTM. The few things I did spot have already been more than covered by Gilles' review so I have nothing more to add really. I have a really small, nit that I don't think is worth iterating again for but I have noted it anyway and leave it up to you whether you think it is worth incorporating.

gilles-peskine-arm reviewed (commented):
Approved after the latest rebase + comment wording improvement.

However, this conflicts with https://github.com/Mbed-TLS/mbedtls/pull/8159 and I'd rather not rebase that a third time in so many days. So please hold off until that is merged, then rebase/merge again.

valeriosetti commented:
@gilles-peskine-arm @tom-daubney-arm I rebased this one after #8159 was merged into `development`.

Note: as in #8033, I didn't change my commits' text compared to the previous status (before the merge of #8159), because I though this could ease your review work. I didn't change the content of the commits indeed, I just moved the changes as necessary.

gilles-peskine-arm reviewed (dismissed):
I'm approving 8c7738bffe31f044d21663e1ece02b8829c9ec4c as a rebased version of my previous approval, with the additions to `config_psa.h` or `build_info.h` now in the correct new file, plus one commit to fix `PSA_WANT` symbols formerly defined without a value.

Regarding the commit messages, they're not ideal, but they're ok: “config_psa section” can be understood to refer to one of the files that `config_psa.h` includes.

gilles-peskine-arm commented on include/mbedtls/config_adjust_legacy_crypto.h:
```diff
@@ -135,6 +135,47 @@
 #define MBEDTLS_PK_PARSE_C
 #endif
 
+/* Helpers to state that each key is supported either on the builtin or PSA side. */
```
Looking again, I'm questioning whether this is the right place to define `MBEDTLS_ECP_HAVE_xxx`. These symbols are similar to `MBEDTLS_MD_CAN_xxx` which are defined in `md.h` and `MBEDTLS_PK_CAN_xxx` which are defined in this file.

Normally code that depends on `MBEDTLS_MD_HAVE_xxx` would have to include `md.h` since it would be calling `mbedtls_md_xxx` functions. However, there could be indirect dependencies where a module A calls another module B that calls md, and A itself doesn't call md functions but needs to have e.g. a buffer size calculation depend on which md algorithms are supported. So there's a risk that we forget to `#include "mbedtls/md.h"` from A and the buffer size is wrong.

Thus I think that we should move the definitions of `MBEDTLS_MD_HAVE_xxx` here.

This is out of scope of this pull request, which is already getting very long. So please don't do it here, let's remember to do it in a follow-up.

mpg replied:
Created https://github.com/Mbed-TLS/mbedtls/issues/8257 to track this, and added in to the "Driver-only cleanups" EPIC (which due to this issue is no longer just about ECC). As a result, marking this converstaas resolved, please shout if you disagree.

gilles-peskine-arm commented on include/mbedtls/config_adjust_psa_superset_legacy.h:
```diff
@@ -70,4 +70,98 @@
+/* Ensure that the PSA's supported curves (PSA_WANT_ECC_xxx) are always a
+ * superset of the builtin ones (MBEDTLS_ECP_DP_xxx). */
+#if defined(MBEDTLS_ECP_DP_BP256R1_ENABLED)
+#define MBEDTLS_SOME_BUILTIN_EC
```
Looking again, this file should not define `MBEDTLS_SOME_BUILTIN_EC`. The purpose of this file is to enable certain PSA mechanisms if the corresponding legacy mechanism is enabled. So it should only have conditionals based on public `MBEDTLS_xxx` symbols (or in a stretch, derived internal `MBEDTLS_xxx` symbols), and it should only define `PSA_WANT_xxx` symbols.

Whether to define `MBEDTLS_SOME_BUILTIN_EC` should only depend on `PSA_WANT_xxx` and `MBEDTLS_PSA_ACCEL_xxx`, regardless of how those were set.

gilles-peskine-arm reviewed (changes requested):
I'm approving the latest rebase and have reviewed up to 3e1fbb9d8e393bfbbcefd08a6665828464defaad.

https://github.com/Mbed-TLS/mbedtls/pull/8075#discussion_r1319639783 has prompted me to check the logic around `MBEDTLS_SOME_BUILTIN_EC` and I think you're right that it was wrong, but your fix isn't right either.

mpg commented:
> * enable builtin algs as long as there is some builtin curve supported and the corresponding PSA_WANT_ALG is enabled (task 8016)

Sorry but that's not correct, and not what the task description said. We want to enable built-in algs when the correspoding `PSA_WANT_ALG` is enabled that the is some curve that is required (`PSA_WANT`) but not accelerated (`MBEDTLS_PSA_ACCEL`).

The condition "required but not accelerated" (RNA) is more restrictive that "built-in" (BI): if there's a curve that's RNA, then it's BI, but the opposite is not true.

I haven't read all the comments yet but I wanted to clarify this first. I see mention of a `MBEDTLS_SOME_BUILTIN_EC` macro and I think that's the wrong macros: what we need is `MBEDTLS_SOME_CURVE_IS_REQUIRED_BUT_NOT_ACCELERATED` (or a shorter name that means the same).

Here's a concrete example where this makes a difference: someone has an friver for ECDH and ECDSA that covers P-256 and P-384, and that's the only curve they want. However they also want EC J-PAKE with P-256 but they don't have an EC J-PAKE driver. So, they'll enable P-256 built-in but they still don't need `ECDH_C` or `ECDSA_C` and we shouldn't force it on them.

The case where we want to auto-enable built-in algs even in the presence of a driver is the following: the user has requested ECDH and ECDSA for P-256 and P-384, and they have a driver that does ECDH and ECDSA but only with P-256. We need to enabled `ECDH_C`, `ECDSA_C` (and their dependencies: `ECP_C` and `BIGNUM_C`) as well as `MBEDTLS_ECP_DP_SECP384R1`, because that's the only way the user will get ECDH and ECDSA with P-384.

So, I think the correct condition is "required but not accelerated" and not "built-in".

mpg commented:
> I haven't read all the comments yet but I wanted to clarify this first.

After reading more of the comments, I realise Gilles made the same point, but better (I forgot about types/actions), in [this comment](https://github.com/Mbed-TLS/mbedtls/pull/8075#discussion_r1319836847). Specifically:

> So `MBEDTLS_SOME_BUILTIN_EC` needs to be defined if there is a curve which is WANT but not ACCEL, **or** there is a type which is WANT but not ACCEL, **or** there is an algorithm/action which is WANT but not ACCEL. All of these tests belong in `config_adjust_legacy_from_psa.h`, at the same time we activate a ` MBEDTLS_PSA_BUILTIN_xxx`.

Except I think in order to reflect this condition the macro name should not be `SOME_BUILTIN_EC` but rather something like `NEED_BUILTIN_EC` or `SOME_EC_NOT_ACCELERATED`.

mpg commented:
We probably need to take a minute to think about the general strategy here. In light of Gilles's comment, I realised things are a bit more complex than I thought, and as a result the issue description is incomplete: it's not just that (1) if some requested curves are not accelerated, we need to enable built-in algs and key types (the current issue description), it's also that (2) if some key types are not accelerated, we need to enable built-in curves and algs, and (3) if some algs are not accelerated, we need to enable built-in curves and key types.

Said otherwise, an ECC feature is defined by a triplet, and if any element of the triplet lacks acceleration, we need to have not only it built-in, but also the other two - so, the whole triplet built-in. I think this actually results in a simple binary:
- either we have driver support for each thing (curve, key type, alg) that's requested, and we don't need to have anything built-in;
- or accelerator support is not complete, and we need to have everything builtin.
- 
In terms for C (pre-processor) code, this could look like:
```c
/* First pass: check completeness of ECC acceleration */

/* Curves */
#if defined(PSA_WANT_ECC_BRAINPOOL_P_R1_256) && !defined(MBEDTLS_PSA_ACCEL_ECC_BRAINPOOL_P_R1_256)
#define MBEDTLS_PSA_ECC_ACCEL_INCOMPLETE
#endif
// [...] (for each ECC curve)

/* Key types */
#if defined(PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_IMPORT) && !defined(MBEDTLS_PSA_ACCEL_KEY_TYPE_ECC_KEY_PAIR_IMPORT)
#define MBEDTLS_PSA_ECC_ACCEL_INCOMPLETE
#endif
// [...] (for each ECC key type)

/* Algs */
#if defined(PSA_WANT_ALG_ECDH) && !defined(MBEDTLS_PSA_ACCEL_ALG_ECDH)
#define MBEDTLS_PSA_ECC_ACCEL_INCOMPLETE
#endif
// [...] (for each ECC alg)

/* Second pass: if anything was missing, we need everything built-in */
#if defined(MBEDTLS_PSA_ECC_ACCEL_INCOMPLETE)
/* Curves */
#if defined(PSA_WANT_ECC_BRAINPOOL_P_R1_256)
#define MBEDTLS_PSA_BUILTIN_ECC_BRAINPOOL_P_R1_256 1
#define MBEDTLS_ECP_DP_BP256R1_ENABLED
#endif /* PSA_WANT_ECC_BRAINPOOL_P_R1_256 */
// [...] (for each ECC curve)

/* Key types */
#if defined(PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_IMPORT)
#define MBEDTLS_PSA_BUILTIN_KEY_TYPE_ECC_KEY_PAIR_IMPORT 1
#define MBEDTLS_ECP_LIGHT
#define MBEDTLS_BIGNUM_C
#endif /* PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_IMPORT */
// [...] (for each ECC key type)

/* Algs */
#if defined(PSA_WANT_ALG_ECDH)
#define MBEDTLS_PSA_BUILTIN_ALG_ECDH 1
#define MBEDTLS_ECDH_C
#define MBEDTLS_ECP_C
#define MBEDTLS_BIGNUM_C
#endif /* !MBEDTLS_PSA_ACCEL_ALG_ECDH */
// [...] (for each ECC alg)

#endif /* MBEDTLS_PSA_ECC_ACCEL_INCOMPLETE */
```

The new macro `MBEDTLS_PSA_ECC_ACCEL_INCOMPLETE` would replace `MBEDTLS_SOME_BUILTIN_EC` which was the wrong notion. It would be defined and used only in `config_adjust_legacy_from_psa.h`.

@gilles-peskine-arm @tom-daubney-arm Wdyt?

gilles-peskine-arm commented:
`MBEDTLS_PSA_ECC_ACCEL_INCOMPLETE` is a better name but the logic isn't right. Let's say P256R1 is accelerated but the driver can only do randomized ECDSA and the user wants deterministic ECDSA. Then we still need to activate the built-in P256R1 implementation and ECDSA implementation.

mpg commented:
> Let's say P256R1 is accelerated but the driver can only do randomized ECDSA and the user wants deterministic ECDSA.

Sure, so `PSA_WANT_ALG_DETERMINISTIC_ECDSA` is defined but `MBEDTLS_PSA_ACCEL_ALG_DETERMINISTIC_ECDSA` is not, so in my example code, this gets detected as part of the fist pass (first "for each alg" block) for each ECC alg)` and causes `MBEDTLS_PSA_ECC_ACCEL_INCOMPLETE` to be defined.

As a consequence, the 2nd pass will not be skipped, and will enable the built-in P256R1 implementation (second "for each curve" block) and ECDSA implementation (2nd "for each alg" block).

What am I missing?

gilles-peskine-arm commented:
Hmm, re-reading, I think I made a mistake when reading your proposal earlier. Now that I look at it again, the logic seems right. Sorry for the noise.

mpg commented:
The update I just pushed does _not_ addressed all know issues and is _not_ ready for re-review, I just wanted some CI feedback.

mpg commented:
Note: the last commit that was reviewed was 3e1fbb9d8e393bfbbcefd08a6665828464defaad, which I saved [here](https://github.com/mpg/mbedtls/tree/8075-1). (I force-pushed yesterday to resolve a conflict, and am force-pushing again today for the same reason.)

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -2930,26 +2958,21 @@ psa_crypto_config_accel_all_curves_except_one () {
+    loc_extra_list="ALG_SHA_1 ALG_SHA_224 ALG_SHA_256 ALG_SHA_384 ALG_SHA_512"
     helper_libtestdriver1_make_drivers "$loc_accel_list" "$loc_extra_list"
 
     # (See above regarding loc_curve_list.)
```
Here and in several other components that are [failing as of c5fa53fc53348571bc3e49503bf530bd31eef5c5](https://mbedtls.trustedfirmware.org//job/mbed-tls-pr-head/job/PR-8075-head/21/flowGraphTable/), I don't understand the logic of the test. Since there is one curve that isn't accelerated, and ECDH (etc.) is requested, `ecdh.c` does get built. The logic of the library makes sense. Why does the test want the built-in ECDH not to be enabled even though it sets up a configuration that requires it?

mpg replied:
Several tests date back from a time when we were not reasoning correctly about acceleration and possibly rely on flawed logic. I'll review them and either fix them, or I guess even remove a few that might not make sense any more. (That was next on my list, and is the main reason I haven't re-added the label "needs-review" yet.)

mpg commented:
> Said otherwise, an ECC feature is defined by a triplet, and if any element of the triplet lacks acceleration, we need to have not only it built-in, but also the other two - so, the whole triplet built-in. I think this actually results in a simple binary:
> 
>     * either we have driver support for each thing (curve, key type, alg) that's requested, and we don't need to have anything built-in;
> 
>     * or accelerator support is not complete, and we need to have everything builtin.

Hmm, not really. I still believe the first part is correct (if any element of a required triplet lacks acceleration, we need all three components of the triplet built-in), but I no longer believe it follows logically that either everything is accelerated, or everything needs to be built-in.

Example:
- we require (`PSA_WANT`): ECDH and ECDSA, all ECC key types, all curves;
- we have acceleration (`MBEDTLS_PSA_ACCEL`) for: all of this except ECDSA;
- as a consequence, we need built-in: ECDSA, all ECC key types, all curves, but not ECDH.

I think the logic is easy to fix by replacing the currently monolithic macro `MBEDTLS_PSA_ECC_ACCEL_INCOMPLETE` with three macros: one for curves, one for key types, one for algs, and then for example an requested alg need to be built-in if either the alg is not accelerate, or curve acceleration is incomplete, or key types acceleration is incomplete.

I'll go ahead and implement that. This should fix at least some of the CI failures.

tom-daubney-arm commented:
Since #8033 is closed in favour of #8075 I have manually linked the PR to make sure that #8014 closes when #8075 merges. Please let me know if this is not desired or incorrect.

gilles-peskine-arm commented on library/ssl_misc.h:
```diff
@@ -2313,20 +2313,20 @@ static inline int mbedtls_ssl_tls13_sig_alg_for_cert_verify_is_supported(
     switch (sig_alg) {
-#if defined(MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED)
-#if defined(PSA_WANT_ALG_SHA_256) && defined(MBEDTLS_ECP_DP_SECP256R1_ENABLED)
+#if defined(MBEDTLS_PK_CAN_ECDSA_SOME)
```
In 9f27b4b379 the guards here change from `defined(MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED` to `defined(MBEDTLS_PK_CAN_ECDSA_SOME)`. The new code looks right, but there's no indication of what's happening in the commit message. `defined(MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED` was added in https://github.com/Mbed-TLS/mbedtls/pull/7921, which just got merged, and this looks like a mistake in #7921. Should we be looking for other mistakes? Why didn't review or testing catch this: is the change from #7921 correct after all?

mpg replied:
Yes, the change was intentional as I noticed the issue while resolving conflicts, and I meant to add a note in the commit message but missed it the first time and forgot to came back for it.

I was planning on creating a follow-up issue for 7921 because I'm not fully happy with the naming scheme for the new macros, I'll either add this to that issue or create two follow-up issues. (I've added this to my personal notes and will create the issue(s) shortly.)

mpg replied:
Aw, the frustrating part is I did notice this during my initial review of #7921, but then incorrectly marked as resolved later... https://github.com/Mbed-TLS/mbedtls/pull/7921/files#r1281542395

mpg replied:
Tracked as #8258 in the cleanups EPIC. Marking this discussion as resolved.

gilles-peskine-arm commented on include/mbedtls/config_psa.h:
```diff
@@ -40,6 +40,10 @@
 
+/* We need this to have a complete list of requirements
+ * before we deduce what built-ins are required. */
+#include "psa/crypto_adjust_keypair_types.h"
```
Not a blocker, but `config_psa.h` is on its way out. It only exists to facilitate the transition with already-written code that isn't merged yet. Please avoid adding new code to it.

mpg replied:
Ok, thanks for the info, I'll keep that in mind!

gilles-peskine-arm commented on include/psa/crypto_adjust_keypair_types.h:
```diff
@@ -0,0 +1,103 @@
+/**
+ * \file psa/crypto_adjust_keypair_types.h
```
Non-blocker: I think the name should be `crypto_adjust_config_key_pair_types.h`. With `config` in the name which I forgot for `crypto_adjust_auto_enabled.h`, and with an underscore in `key_pair` because that's how we spell it in macro names.

gilles-peskine-arm reviewed (commented):
I've reviewed the rebases and a few new commits since [my previous review](https://github.com/Mbed-TLS/mbedtls/pull/8075/files) of 3e1fbb9d8e393bfbbcefd08a6665828464defaad, up to f4b233e04306f9b133efb421550548462e4431cb "Group all ECC-related things in legacy_from_psa.h".

I'll continue reviewing on Friday afternoon, ideally after this passes the CI.

mpg commented:
The only two failures in the last CI run were mistakes in the respective components - so, the components that were passing before should keep passing. There's just the risk of those mistakes hiding other issues in the affected components, but other than that we can have some hope that the CI will come back green this time.

As a consequence, I think it makes sense to review this without waiting for the CI results (considering the time pressure).

mpg commented:
Rebased in order to resolve conflicts after #8240 was merged.

gilles-peskine-arm commented on include/mbedtls/config_adjust_legacy_from_psa.h:
```diff
@@ -157,8 +159,9 @@
+/* Special case: we don't support cooked key derivation in drivers yet */
 #if defined(PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_DERIVE) && \
-    !defined(MBEDTLS_PSA_ACCEL_KEY_TYPE_ECC_KEY_PAIR_DERIVE)
+    (1 || !defined(MBEDTLS_PSA_ACCEL_KEY_TYPE_ECC_KEY_PAIR_DERIVE))
```
Non-blocker: I'd prefer to either comment out the `&& !defined(MBEDTLS_PSA_ACCEL_KEY_TYPE_ECC_KEY_PAIR_DERIVE)` than put a `1 ||` short-circuit. That would make it clearer what part of the expression is a temporary limitation. Alternatively, do `#undef MBEDTLS_PSA_ACCEL_KEY_TYPE_ECC_KEY_PAIR_DERIVE`, which would let us have partial support for `MBEDTLS_PSA_ACCEL_KEY_TYPE_ECC_KEY_PAIR_DERIVE` (e.g. in test code): it just would never get effectively activated as long as we don't add the support. But I can live with `1 ||`.

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -2340,6 +2340,9 @@ component_test_psa_crypto_config_accel_ecdsa () {
     loc_accel_list="ALG_ECDSA ALG_DETERMINISTIC_ECDSA \
                     $(helper_get_psa_key_type_list "ECC")"
 
+    # Note: Those are handled in a special way by the libtestdriver machinery,
```
Non-blocker: I don't really understand this special handling. Why are curves treated differently from key types? But since this is a preexisting problem with the complexity of the libtestdriver1 setup, and we have some checks in place (`grep` in the `all.sh` jobs, driver testing parity checks in outcome analysis), I can live with the changes in this pull requests. I'd probably reject more complex changes without some prior rework of the libtestdriver1 setup.

mpg replied:
I don't understand it either, and I was meaning to create an issue about looking if it can be simplified (once I'm done with release-critical things).

mpg replied:
#8262

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -2987,116 +3041,6 @@ component_test_tfm_config() {
-    psa_crypto_config_accel_all_curves_except_one MBEDTLS_ECP_DP_SECP192R1_ENABLED
-}
-
-component_test_psa_crypto_config_accel_all_curves_except_x25519 () {
```
Blocker: We should have at least one component where at least one curve is accelerated and one curve isn't. (Possibly more due to differences in what you can do with secp256r1 vs other Weierstrass curves vs Montgomery curves.) Unless I'm missing something (which is possible, but then I'd like the commit message for fc455740a84e18f92a575fc16e82f2628afe9561 "Remove components that are no longer needed" to say so), the two `component_test_psa_crypto_config_accel_all_curves_except_xxx` were the only places where this happens. I don't necessarily want those particular two back, but I want something to validate that we haven't accidentally made such a mixed configuration impossible.

mpg replied:
Ah good point, I missed that those components also served this purpose.

Actually, now that you're mentioning this, I don't think we have a test where some key types are accelerated but not all of them. (For example, a driver could support all key operations except generate.) We probably should.

gilles-peskine-arm reviewed (changes requested):
I have finished reviewing the rebases since [my review](https://github.com/Mbed-TLS/mbedtls/pull/8075#pullrequestreview-1611611971) of https://github.com/Mbed-TLS/mbedtls/commit/89cab7987d8c20126e0fb12914dae0839448758c (with intermediate steps at https://github.com/Mbed-TLS/mbedtls/pull/8075#pullrequestreview-1617430822 on https://github.com/Mbed-TLS/mbedtls/commit/3e1fbb9d8e393bfbbcefd08a6665828464defaad and at https://github.com/Mbed-TLS/mbedtls/pull/8075#pullrequestreview-1638566414 on https://github.com/Mbed-TLS/mbedtls/commit/f4b233e04306f9b133efb421550548462e4431cb). I am satisfied with the resolution of https://github.com/Mbed-TLS/mbedtls/pull/8075#discussion_r1319836847 and the subsequent discussion leading to https://github.com/Mbed-TLS/mbedtls/pull/8075#issuecomment-1717235560.

I'm happy with 39de407acb62af726ffaa9bcb59c62facc661502 except that as far as we can tell, we've lost some test coverage, namely of builds with some elliptic curves accelerated and some built in.

daverodgman commented:
@bensze01 looks like another CI issue, this time in `all_u16-check_python_files`:
```

[2023-09-24T16:34:29.801Z] Running mypy ...

[2023-09-24T16:34:29.801Z] /var/lib/builds/.local/lib/python3.5/site-packages/referencing/jsonschema.py:225: error: Variable annotation syntax is only supported in Python 3.6 and greater

[2023-09-24T16:34:29.801Z] Found 1 error in 1 file (checked 32 source files)

[2023-09-24T16:34:29.801Z] ^^^^check_python_files: Lint: Python scripts: tests/scripts/check-python-files.sh -> 1^^^^
```

xkqian commented:
One failure about python check. Not sure why: " * Installed asan versions:

libasan2: 5.4.0-6ubuntu1~16.04.12



  Gen   error.c

  Gen   psa/psa_constant_names_generated.c

  Gen   test/query_config.c

  Gen   version_features.c

  Gen   ssl_debug_helpers_generated.c

  Gen   psa_crypto_driver_wrappers.h psa_crypto_driver_wrappers_no_static.c

  Gen   suites/test_suite_psa_crypto_generate_key.generated.data suites/test_suite_psa_crypto_low_hash.generated.data suites/test_suite_psa_crypto_not_supported.generated.data suites/test_suite_psa_crypto_op_fail.generated.data suites/test_suite_psa_crypto_storage_format.current.data suites/test_suite_psa_crypto_storage_format.v0.data ...

  Gen   psa_crypto_driver_wrappers.h psa_crypto_driver_wrappers_no_static.c

  Gen   visualc/VS2013/mbedTLS.sln ...

  Gen   visualc/VS2013/mbedTLS.vcxproj ...

  Gen   suites/test_suite_ecp.generated.data

  Gen   suites/test_suite_bignum.generated.data suites/test_suite_bignum_core.generated.data suites/test_suite_bignum_mod.generated.data suites/test_suite_bignum_mod_raw.generated.data



******************************************************************

* check_python_files: Lint: Python scripts 

* Mon Sep 25 09:36:00 UTC 2023

******************************************************************

Running pylint ...



------------------------------------

Your code has been rated at 10.00/10





Running mypy ...

/var/lib/builds/.local/lib/python3.5/site-packages/referencing/jsonschema.py:225: error: Variable annotation syntax is only supported in Python 3.6 and greater

Found 1 error in 1 file (checked 32 source files)

^^^^check_python_files: Lint: Python scripts: tests/scripts/check-python-files.sh -> 1^^^^

Is it one known issue? or what else?

gilles-peskine-arm commented:
That's a known issue, should be fixed by https://github.com/Mbed-TLS/mbedtls/pull/8249. See [Slack thread](https://arm-ce.slack.com/archives/C01CC4Y1T1P/p1695629431538349)

gilles-peskine-arm commented on include/mbedtls/config_adjust_legacy_from_psa.h:
```diff
@@ -344,7 +344,6 @@
     defined(MBEDTLS_PSA_ECC_ACCEL_INCOMPLETE_KEYTYPES_BASIC)
 #define MBEDTLS_PSA_BUILTIN_PAKE 1
 #define MBEDTLS_PSA_BUILTIN_ALG_JPAKE 1
-#define MBEDTLS_ECP_DP_SECP256R1_ENABLED
```
The PSA wrapper code hard-codes secp256r1 and SHA-256. It doesn't work with other algorithms. So we do need `MBEDTLS_ECP_DP_SECP256R1_ENABLED`, or more precisely `PSA_WANT_ALG_JPAKE` won't do anything useful without it and `MBEDTLS_MD_CAN_SHA256`.

mpg replied:
Right, I completely misremembered how our current PSA implementation is tied to these algs, sorry.

gilles-peskine-arm reviewed (changes requested):
I have reviewed the rebase + 1 new commit 94ee9c5b6838f1a420dce4270a5230e926acce82 since [my previous review](https://github.com/Mbed-TLS/mbedtls/pull/8075#pullrequestreview-1640612373).

* Ok for the rebase.
* I think the new commit is just making things worse and should be removed.
* The one [blocker](https://github.com/Mbed-TLS/mbedtls/pull/8075#discussion_r1335547715) and several nice-to-have still stand.

mpg commented:
Just force-pushed, removing yesterday's faulty commit and adding two new commits, addressing your blocker comment.

I haven't addressed the non-blockers, and I'm inclined to do them in a follow-up.

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -2520,6 +2520,186 @@ component_test_psa_crypto_config_accel_pake() {
+# - component_test_psa_crypto_config_accel_ecc_weirstrass_curves
+# - component_test_psa_crypto_config_accel_ecc_non_weirstrass_curves
+common_test_psa_crypto_config_accel_ecc_some_curves () {
+    WEIERSTRASS=$1
```
Non-blocker: a common convention in (ba)sh is to use uppercase for environment (exported) variables and lowercase for local variables. Opinions differ on script-global variables that aren't exported.

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -2520,6 +2520,186 @@ component_test_psa_crypto_config_accel_pake() {
+    # helper_get_psa_curve_list that only keeps Weierstrass families.
+    loc_weierstrass_list=""
+    loc_non_weierstrass_list=""
+    for ITEM in $(sed -n 's/^#define PSA_WANT_\(ECC_[0-9A-Z_a-z]*\).*/\1/p' <"$CRYPTO_CONFIG_H"); do
```
Non-blocker: this feels like it's reinventing a wheel that's in `crypto_knowledge.py`.

mpg replied:
I'm not sure how to re-use a wheel from `crypto_knowledge.py` in `all.sh`.

gilles-peskine-arm reviewed (approved):
LGTM

gilles-peskine-arm commented:
The new test components `test_psa_crypto_config_accel_ecc_non_weirstrass_curves` and `test_psa_crypto_config_accel_ecc_weirstrass_curves` are failing. I'm looking into it.

gilles-peskine-arm commented on tests/suites/test_suite_pk.data:
```diff
@@ -13,135 +13,135 @@ depends_on:MBEDTLS_RSA_C:MBEDTLS_GENPRIME
 
 PK utils: ECKEY SECP192R1
-depends_on:MBEDTLS_PK_HAVE_ECC_KEYS:MBEDTLS_ECP_DP_SECP192R1_ENABLED
+depends_on:MBEDTLS_PK_HAVE_ECC_KEYS:MBEDTLS_ECP_HAVE_SECP192R1
```
This test case fails in `all.sh test_psa_crypto_config_accel_ecc_non_weirstrass_curves` in 5c21036d19ff922b65a7fc952f5ea17dafbc3aa5.

If some curves are accelerated but not others, and `MBEDTLS_USE_PSA_CRYPTO` is enabled, we [don't define `MBEDTLS_PK_USE_PSA_EC_DATA`](https://github.com/Mbed-TLS/mbedtls/blob/87fe99627f30983dfdbf7fba0506464853cbf0d2/include/mbedtls/pk.h#L174) (makes sense: that's only for a pure-PSA world, not for a mixed world). Then the code in `test_suite_pk` uses the [legacy code path to generate an EC key](https://github.com/Mbed-TLS/mbedtls/blob/87fe99627f30983dfdbf7fba0506464853cbf0d2/tests/suites/test_suite_pk.function#L104). That's a bug in the test code: nowadays it should base the choice of code path on where the support is for the specific curve, or it could dispatch to PSA whenever PSA is available. I think this violates the objective of https://github.com/Mbed-TLS/mbedtls/pull/8075.

I'm not sure what the best resolution is at this point, given that this PR is already very long and we want to release ASAP. Do we have to fix it now? Or should we tone down the testing here and file it as a bug to be fixed after the release?

tom-daubney-arm replied:
As much as releasing something with a known bug is rather unpleasant, I am inclined to say it's probably the best option here. My vote goes to toning down the testing and fixing with a follow-up PR post release.

daverodgman replied:
Let's spend a bit more time investigating this and seeing if we can get to a point where we can get the testing in good shape. Slipping a few days while we explore this more isn't going to be a big issue, rather that than risk a buggy release.

mpg replied:
I'm inclined to temporarily tone done the testing here and merge, then investigate and fix in a follow-up, just to reduce the risk of conflicts / complexity of rebasing, as well as to unblock #8242 which depends on this one.

Then depending on how the follow-up is going, we can see what the impact on the release is. Wdyt?

tom-daubney-arm replied:
I am in favour of this approach too, personally.

daverodgman replied:
That sounds reasonable to me.

gilles-peskine-arm replied:
Fine with me. That, or keep the offending tests disabled when the curve is accelerated: as far as I can tell, they assume that the key is not opaque (`mbedtls_pk_get_type(&pk) == type`).

I think the tests are attempting to do something reasonable, which isn't working when some curves are accelerated. So I think this indicates that we haven't achieved proper support for builds where some curves are accelerated and others are not. But it's possible that the tests are just overly picky and need to be rewritten, and it's just ok that switching curves to be accelerated is not a transparent change for users of the pk module.

gilles-peskine-arm replied:
My reasoning: starting at 5c21036d19ff922b65a7fc952f5ea17dafbc3aa5, I attempted to make this test pass in `all.sh test_psa_crypto_config_accel_ecc_non_weirstrass_curves`. The test needs to generate an EC key on a curve which may or may not be accelerated. If the curve is not accelerated then `MBEDTLS_ECP_C` is enabled and the `mbedtls_ecp_gen_keypair` code path is taken, which is fine. If all curves are accelerated then `MBEDTLS_PK_USE_PSA_EC_DATA` is enabled, and all is fine. But if this curve is accelerated while others are not, we have a problem. Since `MBEDTLS_ECP_C` is present (for other curves), `MBEDTLS_PK_USE_PSA_EC_DATA` is disabled. But `mbedtls_ecp_gen_keypair` and even `mbedtls_ecp_group_load` won't work for the accelerated curve.

I am worried that we are breaking something that worked in 3.4, however. So *I don't want to just disable the test components*. I'd like to restore the test components that were removed (`
component_test_psa_crypto_config_accel_all_curves_except_p192` and `component_test_psa_crypto_config_accel_all_curves_except_x25519`). If those don't work, that's a regression and I would not approve this pull request.

mpg replied:
Note: there was no `component_test_psa_crypto_config_accel_all_curves_except_p192` (or similar) in 3.4.

Also, this entire PR is fixing an API bug, that was present in 3.4: before this PR, you could declare algs and key types as accelerated but declare no curve as accelerated, and still have your build be driver-only - it was implicitly assumed that curves were accelerated. So, we are breaking things that used to work in 3.4, but that's on purpose, because those should never have worked in the first place. (Now that I'm writing this, I realise we want a ChangeLog entry under "Changes" here.)

> I would much rather sort out some technical debt (in the bridge between pk and psa, and in how we do builds with drivers) before sorting out the problems here. Hence my preference for just shipping what we have (which is, at the very least, useful for builds where everything ECC is accelerated).

I tend to agree that we care mostly about the case where all required ECC is accelerated, and before spending too much time on supporting partially-accelerated we need to think about our priorities.

Of course, if we decide to only support the case where all ECC is accelerated, we should explicitly say so in the ChangeLog and documentation. (Actually, I think that's a bit overkill: we do support the case where some algs are accelerated but not others, what we don't seem to support easily is the case where come curves are accelerated but not all of them.)

gilles-peskine-arm replied:
> there was no component_test_psa_crypto_config_accel_all_curves_except_p192 (or similar) in 3.4.

Ah, ok. So this is (I suspect) a regression compared to development, but not compared to the last release. Then I'm ok with:

* removing the new components that are failing;
* filing an issue describing the problem;
* documenting that we support all-curves-accelerated, but support for some-curves-accelerated is untested and known not to work in all cases.

daverodgman replied:
"some-curves-accelerated is untested and known not to work in all cases" - presumably the case where only SECP256R1 is accelerated is tested and known to work?

gilles-peskine-arm replied:
> the case where only SECP256R1 is accelerated is tested

No, we don't test this anywhere, as far as I can tell. If we did, it would fail for the same reason as `test_psa_crypto_config_accel_ecc_weirstrass_curves` here. In particular, we only test P256-m in the TF-M config, and it only enables that single curve.

daverodgman replied:
Ah, I see - sorry I misunderstood. We support the case where all curves are accelerated, including when that is just one curve: so we support the TF-M use-case.

mpg replied:
> * removing the new components that are failing;
> * documenting that we support all-curves-accelerated, but support for some-curves-accelerated is untested and known not to work in all cases.

Done in [f7dc6cf](https://github.com/Mbed-TLS/mbedtls/pull/8075/commits/f7dc6cfef18627459e2b151c1d1191167ea1c9a6) except I though removing the whole components was a bit heavy-handed: currently we got all of PSA working in those components, that's already something, so I decided to keep it tested and document it as working. Excluded from tests and documented as "not working": PK (because tests fail), and X.509+TLS (because they use PK so it feels wrong to document them as working when PK is not).

> * filing an issue describing the problem;

https://github.com/Mbed-TLS/mbedtls/issues/8255

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -2475,6 +2520,186 @@ component_test_psa_crypto_config_accel_pake() {
+    make test
+}
+
+component_test_psa_crypto_config_accel_ecc_weirstrass_curves () {
```
For future reference:
```suggestion
component_test_psa_crypto_config_accel_ecc_weierstrass_curves () {
```

gilles-peskine-arm commented on tests/scripts/all.sh:
```diff
@@ -2689,7 +2688,8 @@ common_test_psa_crypto_config_accel_ecc_some_curves () {
     msg "test suites: full with accelerated EC algs and $DESC curves"
-    make test
+    # does not work for PK (and above), see #8255
+    make test SKIP_TEST_SUITES=pk,pkparse,pkwrite,x509parse,x509write,ssl,debug
```
I'm willing to approve this as an emergency for the release if we really feel that we can't afford the time of one more CI run, but I'd prefer to set up a configuration without `MBEDTLS_PK_C` or X.509/TLS (i.e. start with `config.py crypo_full; config.py unset MBEDTLS_PK_C` instead of `config.py full`) and then run a plain `make test`.

mpg replied:
Why would you prefer removing things are compile time? Currently everything builds, so why not assert that so that it keeps building in the future?

mpg replied:
Note: in general, I'd really like to merge this ASAP, in order to unblock #8242. That doesn't prevent us from getting a follow-up PR in the release with the subset of fixes we care about (and possibly more depending on how other release-critical things are progressing).

gilles-peskine-arm replied:
> Why would you prefer removing things are compile time?

Because “it builds but we don't know what it does” is worse than “it's unsupported, don't do this”. I'd even prefer to have an `#error` than mysterious errors that may happen either at compile time or run time.

daverodgman replied:
It sounds like we should keep this PR as-is, to make progress, but in parallel prepare a PR to add a test as Gilles suggests?

mpg replied:
>Because “it builds but we don't know what it does” is worse than “it's unsupported, don't do this”.

The documentation doesn't say it builds, it says "there are known issues in PK, and X.509 and TLS are untested" which I think is much closer to "it's unsupported, don't do this".

> I'd even prefer to have an #error than mysterious errors that may happen either at compile time or run time.

Now, that's an idea I like: not everyone reads the documentation, so an `#error` is safer. But then it goes beyond your initial comment. I can see the point of not building if we take advantage of it in order to give an `#error`.

gilles-peskine-arm reviewed (approved):
As discussed in https://github.com/Mbed-TLS/mbedtls/pull/8075#discussion_r1337144555 and https://github.com/Mbed-TLS/mbedtls/pull/8075/files/5c21036d19ff922b65a7fc952f5ea17dafbc3aa5..bdc678f9101a814782ec10de6ed14f7ac5c1cb1f#r1338309873, we have a test gap in configurations where ECC support is partly accelerated and partly built in. This is a plausible configuration — for example, I want constant-time x25519 and ecdsa on secp256r1, so I enable p256-m and Everest and disable the rest. I am satisfied that we do document it as officially unsupported, but that's only in a file where people would not naturally look. We are not sure what happens in this test gap: maybe it's just problems in our tests, maybe there's an architectural problem in the library and some plausible applications work while others fail with non-security errors, maybe there's a mistake in some buffer calculation that we haven't identified …

As a consequence, **I am approving this pull request** as of  with a reservation: before the release, I want to **take reasonable steps to prevent building suspicious configurations**. I don't want a 100% guarantee that all broken configurations are detected at compile-time, but I want at least to have some validation that a plausible build either is rejected or has passing tests.
